### PR TITLE
Namegen.next_ident_away_in_goal: don't avoid names of nonlocal refs

### PR DESCRIPTION
**Pull request by SkySkimmer**
Fix #3523

Overlays:
- https://github.com/mit-pdos/argosy/pull/4
- https://github.com/fblanqui/color/pull/36
- https://github.com/DeepSpec/sfdev/pull/401
- https://github.com/PrincetonUniversity/VST/pull/528
- https://github.com/MetaCoq/metacoq/pull/603
- TODO

### CODE REVIEW

**Comment by SkySkimmer** [comment]
Overlays aren't too bad, the sf one was pretty annoying because of that `Definition x` though https://github.com/DeepSpec/sf/blob/84060de4a7583c54fd54b263ae867beee2d475ea/plf-current/Stlc.v#L196

**Comment by ppedrot** [comment]
@SkySkimmer I don't think we should rush this one for 8.15, ideally we should replace `next_ident_away_in_goal` by more robust functions in some cases rather than replacing it globally. I had a look at it and some uses in tactics are definitely weird. If so it'd be better to break things only once rather than several times.

**Comment by SkySkimmer** [comment]
>@SkySkimmer I don't think we should rush this one for 8.15, ideally we should replace next_ident_away_in_goal by more robust functions in some cases rather than replacing it globally. I had a look at it and some uses in tactics are definitely weird. If so it'd be better to break things only once rather than several times.

I didn't notice any weird changes while doing overlays

**Comment by ppedrot** [comment]
Weird as in "why the hell do we care about the name being different from an imported one in the context". This makes sense for detyping where this is going to be shown to the user, but some callers are likely to be insensitive to this particular choice. The reason why I want to get rid of this is that relying on the nametab is broken if we ever manage to get rid of the imperative implementation of side-effects.

**Comment by SkySkimmer** [comment]
From doing the stdlib adaptation of the initial version of this PR, it's because the shadowing is likely to force annoying quantifications to refer to the constant.

**Comment by SkySkimmer** [comment]
The point of this PR is that it's not imported names anymore.
We still avoid local names because doing otherwise would break too much, and also it seems to lead to annoying quantifications (I guess you're more likely to refer to local names than random imported ones).
Not sure what this has to do with sde effects.

**Comment by ppedrot** [comment]
Side-effects modify the nametab.

File: doc/changelog/04-tactics/15071-namegen-no-glob.rst [hunk 1]
```diff
@@ -0,0 +1,5 @@
+- **Changed:** Name generation does not avoid names of references
+  except for those from the current module and its ancestors (`#15071
+  <https://github.com/coq/coq/pull/15071>`_, fixes `#3523
+  <https://github.com/coq/coq/issues/3523>`_, by Gaëtan Gilbert).
+  The old behaviour may be restored using :flag:`Avoid Local Imported Names`.
```
Comment (jfehrle) [inline doc/changelog/04-tactics/15071-namegen-no-glob.rst:5]:
```suggestion
- **Changed:** Name generation no longer avoids names of references
  except for those from the current module and its ancestors (`#15071
  <https://github.com/coq/coq/pull/15071>`_, fixes `#3523
  <https://github.com/coq/coq/issues/3523>`_, by Gaëtan Gilbert).
  The old behavior can be restored using :flag:`Avoid Local Imported Names`.
```

"Avoid Local Imported Names" is a bit of a mouthful but I have no better suggestion.

File: doc/sphinx/proof-engine/tactics.rst [hunk 2]
```diff
@@ -1021,6 +1021,21 @@ Managing the local context
...
+      inductives, constructors) from the current open module and its
+      ancestors so that those objects may be referred to without
+      awkward quantification.
+
+      When this flag is on (it is off by default), imported names from
+      modules defined in the current file are also avoided.
+
+      .. deprecated:: 8.14
+
+         This flag is provided for compatibility with versions previous to 8.15.
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/tactics.rst:1037]:
```suggestion
         This flag is provided for compatibility with versions prior to 8.15.
```

File: doc/sphinx/proof-engine/tactics.rst [hunk 3]
```diff
@@ -1021,6 +1021,21 @@ Managing the local context
...
+      Name generation for names introduced by :tacn:`intro` and other
+      name introducing tactics avoids the names of objects (constants,
+      inductives, constructors) from the current open module and its
+      ancestors so that those objects may be referred to without
+      awkward quantification.
+
+      When this flag is on (it is off by default), imported names from
+      modules defined in the current file are also avoided.
+
+      .. deprecated:: 8.14
```
Comment (jfehrle) [inline doc/sphinx/proof-engine/tactics.rst:1035]: Should be at least 8.15, right?  You plan to remove the flag soon, e.g. in 8.16?

**Review by jfehrle — COMMENTED** [review]
A few tweaks.

Also, not so sure this is a great idea--it adds more work for users but I don't see how it benefits them.

File: doc/sphinx/proof-engine/tactics.rst (hunk 3 above)
Comment (SkySkimmer) [reply doc/sphinx/proof-engine/tactics.rst:1035]: >Should be at least 8.15, right? 

yes

> You plan to remove the flag soon, e.g. in 8.16?

no

File: doc/sphinx/proof-engine/tactics.rst [hunk 4]
```diff
@@ -1021,6 +1021,21 @@ Managing the local context
       the latter can still be referred to by a qualified name
       (see :ref:`Qualified-names`).
 
+   .. flag:: Avoid Local Imported Names
+
+      Name generation for names introduced by :tacn:`intro` and other
+      name introducing tactics avoids the names of objects (constants,
+      inductives, constructors) from the current open module and its
+      ancestors so that those objects may be referred to without
+      awkward quantification.
```
[the remark this answers is not in the document]
Comment (SkySkimmer) [reply doc/sphinx/proof-engine/tactics.rst:1030]: instead I'm putting the part that doesn't change before the flag directive

**Comment by SkySkimmer** [comment]
>Also, not so sure this is a great idea--it adds more work for users but I don't see how it benefits them.

The generated names are better and easier to predict IMO

**Comment by ppedrot** [comment]
@SkySkimmer status of this?

**Pull request closed without merging**